### ISSUE TITLE
`make_folds` in `LmtpTask` reads non-existent column `._lmtp_id`, silently disabling clustered cross-fitting folds when `id=` is supplied

### ISSUE THREAD

**Issue by nt-williams** [nt-williams/lmtp#188]
From Molly Davies:

In `R/Task.R`, the private method `make_folds` reads `self$natural$._lmtp_id`, but the column that is actually written by `as_lmtp_data` (and used everywhere else in the package) is named `..i..lmtp_id`. Because reading a non-existent column from a data frame returns `NULL`, the conditional in `make_folds` always falls into the `else` branch with `cluster_ids = NULL`. The result is that **whenever a user passes `id=` to `lmtp_tmle` / `lmtp_sdr` / `lmtp_sub` / `lmtp_ipw`, clustered cross-fitting folds are silently disabled** and the package returns row-level V-fold CV instead. As a secondary effect, the intended stratified-by-outcome branch for non-clustered binomial outcomes is also unreachable.

This is a silent statistical bug, not a runtime error, so it would not be caught by `R CMD check` nor by the existing test suite.

## Root cause

In [`R/Task.R`](https://github.com/nt-williams/lmtp/blob/master/R/Task.R) on `master` and in CRAN 1.5.4:

```r

# Task.R, line 140 (inside private$make_folds)

id <- self$natural$._lmtp_id        # <-- reads a non-existent column, returns NULL


# Task.R, line 167 (inside private$as_lmtp_data)

data$..i..lmtp_id <- self$id        # <-- the column is actually called ..i..lmtp_id

```

Every other site in the package that touches this column uses the correct name `..i..lmtp_id` (see `R/density_ratios.R`, `R/sdr.R`, `R/tmle.R`).

Because `self$natural$._lmtp_id` is `NULL`, the conditional

```r

if (length(unique(id)) == self$n & self$outcome_type == "binomial") {

  ...

} else {

  folds <- make_folds(self$natural, cluster_ids = id, V = V)   # id is NULL here

}

```

always takes the `else` branch with `cluster_ids = NULL`, and `R/make_folds.R` then degenerates to row-level V-fold CV.

## Reproducible example

```r

suppressPackageStartupMessages({

  library(lmtp)

  library(SuperLearner)

})


set.seed(1)

n_clusters <- 20

obs_per_cluster <- 10

n <- n_clusters * obs_per_cluster

cluster_id <- rep(seq_len(n_clusters), each = obs_per_cluster)


W <- rnorm(n); A <- W + rnorm(n); Y <- 0.5 * A + W + rnorm(n)

dat <- data.frame(W = W, A = A, Y = Y, cluster_id = cluster_id)

policy <- function(data, x) data[[x]] - 1


# Build the task exactly as the estimators do internally.

# (LmtpTask is not exported, hence the triple-colon; the bug lives inside

# this class, so inspecting it directly is the cleanest demonstration.)

shifted <- lmtp:::make_shifted(dat, "A", NULL, policy, NULL)

task <- lmtp:::LmtpTask$new(

  data = dat, shifted = shifted,

  A = "A", Y = "Y", L = NULL, W = "W", C = NULL, D = NULL,

  k = Inf, id = "cluster_id", outcome_type = "continuous",

  bounds = NULL, folds = 5, weights = NULL

)


# 1. The column-name typo, observed directly:

is.null(task$natural$._lmtp_id)                              # TRUE  (column doesn't exist)

identical(task$natural$..i..lmtp_id, task$id)                # TRUE  (real column is fine)


# 2. The downstream effect: task$folds splits clusters across train/val.

count_split_clusters <- function(folds, cid) {

  vapply(folds, function(f) {

    trc <- unique(cid[f$training_set])

    vlc <- unique(cid[f$validation_set])

    length(intersect(trc, vlc))

  }, integer(1))

}

count_split_clusters(task$folds, cluster_id)

# [1] 18 18 17 17 16   <-- of 20 clusters, almost all split across folds


# 3. Calling the helper directly with the correct argument gives clustered folds:

correct <- lmtp:::make_folds(task$natural, V = 5, cluster_ids = task$id)

count_split_clusters(correct, cluster_id)

# [1] 0 0 0 0 0


# 4. Sanity check via the public API: lmtp_tmle exhibits the same bug because it

#    uses LmtpTask$make_folds under the hood.

fit <- lmtp_tmle(

  data = dat, trt = "A", outcome = "Y", baseline = "W",

  shift = policy, outcome_type = "continuous",

  id = "cluster_id", folds = 2,

  learners_outcome = "SL.glm", learners_trt = "SL.glm",

  mtp = TRUE,

  control = lmtp_control(.learners_outcome_folds = 2, .learners_trt_folds = 2)

)

# Returns an estimate (no error), but cross-fitting folds were not clustered.

```

### Output observed (lmtp 1.5.4, R 4.3.3)

```

--- Bug evidence (column-name typo in Task.R::make_folds) ---

Task.R:140 reads self$natural$._lmtp_id; that column does NOT exist.

  is.null(task$natural$._lmtp_id) ...........  TRUE

Task.R:167 actually writes column ..i..lmtp_id; that column IS correct.

  identical(task$natural$..i..lmtp_id, task$id)  TRUE


--- Per-fold cluster leakage in BROKEN folds (task$folds) ---

(For clustered CV every value below should be 0.)

  Fold 1: clusters appearing in both train and val = 18 / 20

  Fold 2: clusters appearing in both train and val = 18 / 20

  Fold 3: clusters appearing in both train and val = 17 / 20

  Fold 4: clusters appearing in both train and val = 17 / 20

  Fold 5: clusters appearing in both train and val = 16 / 20


--- Per-fold cluster leakage in CORRECT folds ---

(lmtp:::make_folds(task$natural, V = 5, cluster_ids = task$id))

  Fold 1: clusters appearing in both train and val = 0 / 20

  Fold 2: clusters appearing in both train and val = 0 / 20

  Fold 3: clusters appearing in both train and val = 0 / 20

  Fold 4: clusters appearing in both train and val = 0 / 20

  Fold 5: clusters appearing in both train and val = 0 / 20

```

## Expected vs. observed

- **Expected:** when `id="cluster_id"` is supplied, every cluster appears in exactly one fold's validation set; no cluster has rows in both a fold's training set and its validation set.

- **Observed:** with 20 clusters of 10 observations and 5 folds, 16–18 of 20 clusters are split across train and validation in every fold. The cluster structure is effectively ignored for cross-fitting.

## Impact

- **Primary:** clustered cross-fitting is silently disabled whenever `id=` is provided. SuperLearner's internal cross-validated risk estimates and the cross-fitted nuisance estimators in TMLE/SDR/sub/IPW therefore leak information across clusters, which is exactly what cluster-CV is meant to prevent. This is the explicit motivation of #55 and #57 (the introduction of `id=`).

- **Secondary:** the intended stratified-by-outcome folds for non-clustered binomial outcomes also never run, because the same condition guarding that branch is never reached.

- **Standard errors:** cluster-aware standard errors in `lmtp_contrast` (#110, fixed in #111) rely on `fit$id`, not on the folds, and are therefore unaffected. The point-estimate cross-fitting procedure is the affected component.

This issue is distinct from #105 (which described `id=` *erroring out* in an earlier version), #110/#111 (which fixed clustered SEs in `lmtp_contrast`), and #55/#57 (which originally introduced clustering support).

## Suggested fix

Minimal one-character correction in [`R/Task.R`](https://github.com/nt-williams/lmtp/blob/master/R/Task.R#L140):

```diff

     make_folds = function(V) {

-      id <- self$natural$._lmtp_id

+      id <- self$natural$..i..lmtp_id

```

A self-contained patch with this fix plus a regression test for `tests/testthat/test-Task.R` is attached at the bottom of this issue. Optionally, the branching condition could be tightened to make intent explicit (e.g. `if (is.null(self$id_var)) { stratified branch } else { clustered branch }`), but the one-character change above is sufficient to restore the originally intended behavior.

## Suggested regression test

```r

test_that("task folds respect cluster ids when id= is supplied", {

  set.seed(1)

  n_clusters <- 20

  cid <- rep(seq_len(n_clusters), each = 10)

  n <- length(cid)

  d <- data.frame(W = rnorm(n), A = rnorm(n), Y = rnorm(n), cluster_id = cid)

  sh <- make_shifted(d, "A", NULL, function(data, x) data[[x]] - 1, NULL)


  task <- LmtpTask$new(

    data = d, shifted = sh,

    A = "A", Y = "Y", L = NULL, W = "W", C = NULL, D = NULL,

    k = Inf, id = "cluster_id", outcome_type = "continuous",

    bounds = NULL, folds = 5, weights = NULL

  )


  for (f in task$folds) {

    train_clusters <- unique(cid[f$training_set])

    val_clusters   <- unique(cid[f$validation_set])

    expect_length(intersect(train_clusters, val_clusters), 0)

  }

})

```

## Patch (`git apply` from the package root, i.e. the directory containing `DESCRIPTION`)

```diff

diff --git a/R/Task.R b/R/Task.R

--- a/R/Task.R

+++ b/R/Task.R

@@ -137,7 +137,7 @@ LmtpTask <- R6::R6Class(

     },

 

     make_folds = function(V) {

-      id <- self$natural$._lmtp_id

+      id <- self$natural$..i..lmtp_id

 

       if (length(unique(id)) == self$n & self$outcome_type == "binomial") {

         strata <- self$natural[[final_outcome(self$vars$Y)]]

diff --git a/tests/testthat/test-Task.R b/tests/testthat/test-Task.R

--- a/tests/testthat/test-Task.R

+++ b/tests/testthat/test-Task.R

@@ -50,3 +50,28 @@ test_that("task creation with survival outcome", {

   expect_equal(head(task$is_outcome_free(sim_point_surv, 5)), c(F, F, F, T, F, F))

   expect_equal(head(task$is_competing_risk_free(sim_point_surv, 5)), rep(TRUE, 6))

 })

+

+test_that("task folds respect cluster ids when id= is supplied", {

+  set.seed(1)

+  n_clusters <- 20

+  cid <- rep(seq_len(n_clusters), each = 10)

+  n <- length(cid)

+  d <- data.frame(W = rnorm(n), A = rnorm(n), Y = rnorm(n), cluster_id = cid)

+  sh <- make_shifted(d, "A", NULL, function(data, x) data[[x]] - 1, NULL)

+

+  task <- LmtpTask$new(

+    data = d, shifted = sh,

+    A = "A", Y = "Y", L = NULL, W = "W", C = NULL, D = NULL,

+    k = Inf, id = "cluster_id", outcome_type = "continuous",

+    bounds = NULL, folds = 5, weights = NULL

+  )

+

+  # Every cluster must live entirely in one fold's validation set; no overlap

+  # with that fold's training set.

+  for (f in task$folds) {

+    train_clusters <- unique(cid[f$training_set])

+    val_clusters   <- unique(cid[f$validation_set])

+    expect_length(intersect(train_clusters, val_clusters), 0)

+  }

+})

```

## Versions

Tested against: `lmtp` 1.5.4 (CRAN, 2026-05-07) and current `master` at `[github.com/nt-williams/lmtp`](http://github.com/nt-williams/lmtp) — the typo on `R/Task.R:140` is identical in both.

<details>

<summary><code>sessionInfo()</code> from a fresh R session</summary>

```

R version 4.3.3 (2024-02-29)

Platform: x86_64-pc-linux-gnu (64-bit)

Running under: Ubuntu 24.04.2 LTS


Matrix products: default

BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.12.0

LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.12.0


locale:

 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8

 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8

 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C

[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C


time zone: UTC

tzcode source: system (glibc)


attached base packages:

[1] splines   stats     graphics  grDevices utils     datasets  methods

[8] base


other attached packages:

[1] R6_2.6.1            SuperLearner_2.0-40 gam_1.22-7

[4] foreach_1.5.2       nnls_1.6            lmtp_1.5.4


loaded via a namespace (and not attached):

[1] compiler_4.3.3      backports_1.5.1     generics_0.1.4

[4] codetools_0.2-19    iterators_1.0.14    data.table_1.18.2.1

[7] checkmate_2.3.4

```

</details>

**Comment by nt-williams** [comment]
Patched in devel
